clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,330 @@
+            } else {
+                onFailure(exception, response.receivedTimeMs());
+            }
+        });
+        return request;
+    }
+
+    private void onFailure(final Throwable exception, final long responseTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
+        logger.warn("Failed to send heartbeat to coordinator node {} due to error: {}",
+                coordinatorRequestManager.coordinator(), exception.getMessage());
+    }
+
+    private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
+        if (Errors.forCode(response.data().errorCode()) == Errors.NONE) {
+            this.heartbeatRequestState.updateHeartbeatIntervalMs(response.data().heartbeatIntervalMs());
+            this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
+            this.heartbeatRequestState.resetTimer();
+            this.membershipManager.updateState(response.data());
+            return;
+        }
+        onErrorResponse(response, currentTimeMs);
+    }
+
+    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
+                                 final long currentTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        Errors error = Errors.forCode(response.data().errorCode());
+        maybeTransitionToFailureState(error);
+        String errorMessage = response.data().errorMessage();
+        switch(error) {
+            case NOT_COORDINATOR:
+            case COORDINATOR_NOT_AVAILABLE:
+                String message = String.format("Coordinator node %s is either not started or not valid. Retrying",
+                    coordinatorRequestManager.coordinator());
+                logInfo(errorMessage, response, currentTimeMs);
+                coordinatorRequestManager.markCoordinatorUnknown(response.data().errorMessage(), currentTimeMs);
+                break;
+            case COORDINATOR_LOAD_IN_PROGRESS:
+                // retry
+                message = String.format("Heartbeat was not successful because the coordinator node %s is " +
+                    "loading."
+                    + "Retrying", coordinatorRequestManager.coordinator());
+                logInfo(errorMessage, response, currentTimeMs);
+                break;
+            case GROUP_AUTHORIZATION_FAILED:
+                GroupAuthorizationException exception =
+                    GroupAuthorizationException.forGroupId(membershipManager.groupId());
+                logger.error("GroupHeartbeatRequest failed due to group authorization failure: {}", exception.getMessage());
+                nonRetriableErrorHandler.handle(exception, errorMessage);
+                break;
+            case INVALID_REQUEST:
+                logger.error("GroupHeartbeatRequest failed due to invalid request error: {}", errorMessage);
+                nonRetriableErrorHandler.handle(Errors.INVALID_REQUEST.exception(), errorMessage);
```

Comment:
I don't really understand how the error message is handled here. Is it going to be added as the message of the exception later on? When I mentioned this in my earlier comment, I means doing this `Errors.INVALID_REQUEST.exception(errorMessage)`. This uses the provided error message.